File: SimpleMoxieSwitcher-Windows/SimpleMoxieSwitcher/ViewModels/ContentViewModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using System.Net.Http;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;
using System.Windows.Threading;
using SimpleMoxieSwitcher.Models;
using SimpleMoxieSwitcher.Services;

namespace SimpleMoxieSwitcher.ViewModels
{
    /// <summary>
    /// Main ViewModel for the application content and navigation
    /// </summary>
    public class ContentViewModel : INotifyPropertyChanged
    {
        private bool _isUpdating;
        private string? _statusMessage;
        private ObservableCollection<Personality> _allPersonalities = new();
        private bool _showCustomCreator;
        private bool _showAppearance;
        private bool _showConversations;
        private bool _showSettings;
        private Personality? _editingPersonality;
        private bool _showPersonalityEditor;
        private bool _showChat;
        private bool _showStoryTime;
        private bool _showLearning;
        private bool _showLanguage;
        private bool _showMusic;
        private bool _showSmartHome;
        private bool _showPuppetMode;
        private bool _showChildProfile;
        private bool _showGames;
        private bool _showDocumentation;
        private bool _showModelSelector;
        private Personality? _selectedPersonalityForChat;
        private bool _showSetupWizard;
        private bool _isStartingDocker;
        private bool _isOnline;
        private DateTime? _onlineTime;
        private ObservableCollection<LearningTile> _learningTiles = new();
        private ObservableCollection<StoryTile> _storyTiles = new();

        private readonly IPersonalityService _personalityService;
        private readonly DispatcherTimer _statusCheckTimer;
        private readonly HttpClient _httpClient;

        public ContentViewModel(IPersonalityService personalityService)
        {
            _personalityService = personalityService;
            _httpClient = new HttpClient { Timeout = TimeSpan.FromSeconds(5) };

            LoadPersonalities();
            LoadTiles();
            CheckFirstLaunch();

            // Start status monitoring
            _statusCheckTimer = new DispatcherTimer
            {
                Interval = TimeSpan.FromSeconds(5)
            };
            _statusCheckTimer.Tick += async (s, e) => await CheckMoxieStatusAsync();
            _statusCheckTimer.Start();

            // Check immediately
            _ = CheckMoxieStatusAsync();
        }

        #region Properties

        public bool IsUpdating
        {
            get => _isUpdating;
            set => SetProperty(ref _isUpdating, value);
        }

        public string? StatusMessage
        {
            get => _statusMessage;
            set => SetProperty(ref _statusMessage, value);
        }

        public ObservableCollection<Personality> AllPersonalities
        {
            get => _allPersonalities;
            set => SetProperty(ref _allPersonalities, value);
        }

        public bool ShowCustomCreator
        {
            get => _showCustomCreator;
            set => SetProperty(ref _showCustomCreator, value);
        }

        public bool ShowAppearance
        {
            get => _showAppearance;
            set => SetProperty(ref _showAppearance, value);
        }

        public bool ShowConversations
        {
            get => _showConversations;
            set => SetProperty(ref _showConversations, value);
        }

        public bool ShowSettings
        {
            get => _showSettings;
            set => SetProperty(ref _showSettings, value);
        }

        public Personality? EditingPersonality
        {
            get => _editingPersonality;
            set => SetProperty(ref _editingPersonality, value);
        }

        public bool ShowPersonalityEditor
        {
            get => _showPersonalityEditor;
            set => SetProperty(ref _showPersonalityEditor, value);
        }

        public bool ShowChat
        {
            get => _showChat;
            set => SetProperty(ref _showChat, value);
        }

        public bool ShowStoryTime
        {
            get => _showStoryTime;
            set => SetProperty(ref _showStoryTime, value);
        }

        public bool ShowLearning
        {
            get => _showLearning;
            set => SetProperty(ref _showLearning, value);
        }

        public bool ShowLanguage
        {
            get => _showLanguage;
            set => SetProperty(ref _showLanguage, value);
        }

        public bool ShowMusic
        {
            get => _showMusic;
            set => SetProperty(ref _showMusic, value);
        }

        public bool ShowSmartHome
        {
            get => _showSmartHome;
            set => SetProperty(ref _showSmartHome, value);
        }

        public bool ShowPuppetMode
        {
            get => _showPuppetMode;
            set => SetProperty(ref _showPuppetMode, value);
        }

        public bool ShowChildProfile
        {
            get => _showChildProfile;
            set => SetProperty(ref _showChildProfile, value);
        }

        public bool ShowGames
        {
            get => _showGames;
            set => SetProperty(ref _showGames, value);
        }

        public bool ShowDocumentation
        {
            get => _showDocumentation;
            set => SetProperty(ref _showDocumentation, value);
        }

        public bool ShowModelSelector
        {
            get => _showModelSelector;
            set => SetProperty(ref _showModelSelector, value);
        }

        public Personality? SelectedPersonalityForChat
        {
            get => _selectedPersonalityForChat;
            set => SetProperty(ref _selectedPersonalityForChat, value);
        }

        public bool ShowSetupWizard
        {
            get => _showSetupWizard;
            set => SetProperty(ref _showSetupWizard, value);
        }

        public bool IsStartingDocker
        {
            get => _isStartingDocker;
            set => SetProperty(ref _isStartingDocker, value);
        }

        public bool IsOnline
        {
            get => _isOnline;
            set => SetProperty(ref _isOnline, value);
        }

        public DateTime? OnlineTime
        {
            get => _onlineTime;
            set => SetProperty(ref _onlineTime, value);
        }

        public ObservableCollection<LearningTile> LearningTiles
        {
            get => _learningTiles;
            set => SetProperty(ref _learningTiles, value);
        }

        public ObservableCollection<StoryTile> StoryTiles
        {
            get => _storyTiles;
            set => SetProperty(ref _storyTiles, value);
        }

        #endregion

        #region Methods

        public void CheckFirstLaunch()
        {
            var hasCompletedSetup = Properties.Settings.Default.HasCompletedSetup;
            if (!hasCompletedSetup)
            {
                ShowSetupWizard = true;
            }
        }

        public async Task CheckMoxieStatusAsync()
        {
            var endpoint = Properties.Settings.Default.MoxieEndpoint;
            if (string.IsNullOrEmpty(endpoint))
            {
                endpoint = "http://localhost:8003/hive/endpoint/";
            }

            if (!Uri.TryCreate(endpoint, UriKind.Absolute, out var uri))
            {
                IsOnline = false;
                OnlineTime = null;
                return;
            }

            try
            {
                var response = await _httpClient.GetAsync(uri);
                if (response.IsSuccessStatusCode)
                {
                    if (!IsOnline)
                    {
                        // Just came online
                        OnlineTime = DateTime.Now;
                    }
                    IsOnline = true;
                }
                else
                {
                    IsOnline = false;
                    OnlineTime = null;
                }
            }
            catch
            {
                IsOnline = false;
                OnlineTime = null;
            }
        }

        public void LoadPersonalities()
        {
            var builtInPersonalities = Personality.GetAllPersonalities();
            AllPersonalities = new ObservableCollection<Personality>(builtInPersonalities);
        }

        public void LoadTiles()
        {
            // TODO: Load from database or file system
            LearningTiles = new ObservableCollection<LearningTile>();
            StoryTiles = new ObservableCollection<StoryTile>();
        }

        public async Task SwitchPersonalityAsync(Personality personality)
        {
            IsUpdating = true;
            StatusMessage = $"Switching to {personality.Name}...";

            try
            {
                await _personalityService.SwitchPersonalityAsync(personality);
                StatusMessage = $"✅ SUCCESS! Moxie is now {personality.Emoji} {personality.Name}!";

                // Clear status after delay
                await Task.Delay(3000);
                StatusMessage = null;
            }
            catch (Exception ex)
            {
                StatusMessage = $"❌ Error: {ex.Message}";
            }

            IsUpdating = false;
        }

        public void EditPersonality(Personality personality)
        {
            EditingPersonality = personality;
            ShowPersonalityEditor = true;
        }

        public void OpenChat(Personality personality)
        {
            SelectedPersonalityForChat = personality;
            ShowChat = true;
        }

        public async Task StartDockerContainerAsync()
        {
            IsStartingDocker = true;
            StatusMessage = "Starting Docker container...";

            try
            {
                // Use Docker CLI to start container
                var process = new System.Diagnostics.Process
                {
                    StartInfo = new System.Diagnostics.ProcessStartInfo
                    {
                        FileName = "docker",
                        Arguments = "start openmoxie",
                        RedirectStandardOutput = true,
                        RedirectStandardError = true,
                        UseShellExecute = false,
                        CreateNoWindow = true
                    }
                };

                process.Start();
                var output = await process.StandardOutput.ReadToEndAsync();
                var error = await process.StandardError.ReadToEndAsync();
                await process.WaitForExitAsync();

                if (process.ExitCode == 0)
                {
                    StatusMessage = "✅ Docker container started successfully!";
                    await CheckMoxieStatusAsync();
                }
                else
                {
                    // Try running a new container
                    var runProcess = new System.Diagnostics.Process
                    {
                        StartInfo = new System.Diagnostics.ProcessStartInfo
                        {
                            FileName = "docker",
                            Arguments = "run -d --name openmoxie -p 8003:8003 embodied/openmoxie:latest",
                            RedirectStandardOutput = true,
                            RedirectStandardError = true,
                            UseShellExecute = false,
                            CreateNoWindow = true
                        }
                    };

                    runProcess.Start();
                    await runProcess.WaitForExitAsync();

                    if (runProcess.ExitCode == 0)
                    {
                        StatusMessage = "✅ Docker container created and started successfully!";
                        await CheckMoxieStatusAsync();
                    }
                    else
                    {
                        StatusMessage = $"❌ Failed to start Docker: {error}";
                    }
                }
            }
            catch (Exception ex)
            {
                StatusMessage = $"❌ Error starting Docker: {ex.Message}";
            }

            // Clear status after delay
            await Task.Delay(3000);
            StatusMessage = null;
            IsStartingDocker = false;
        }

        #endregion

        #region INotifyPropertyChanged

        public event PropertyChangedEventHandler? PropertyChanged;

        protected virtual void OnPropertyChanged([CallerMemberName] string? propertyName = null)
        {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        }

        protected bool SetProperty<T>(ref T field, T value, [CallerMemberName] string? propertyName = null)
        {
            if (EqualityComparer<T>.Default.Equals(field, value))
                return false;

            field = value;
            OnPropertyChanged(propertyName);
            return true;
        }

        #endregion
    }
}

File: SimpleMoxieSwitcher/Sources/SimpleMoxieSwitcher/ViewModels/ContentViewModel.swift
import Foundation
import SwiftUI

@MainActor
final class ContentViewModel: ObservableObject {
    @Published var isUpdating = false
    @Published var statusMessage: String?
    @Published var allPersonalities: [Personality] = []
    @Published var showCustomCreator = false
    @Published var showAppearance = false
    @Published var showConversations = false
    @Published var showSettings = false
    @Published var editingPersonality: Personality?
    @Published var showPersonalityEditor = false
    @Published var showChat = false
    @Published var showStoryTime = false
    @Published var showLearning = false
    @Published var showLanguage = false
    @Published var showMusic = false
    @Published var showSmartHome = false
    @Published var showPuppetMode = false
    @Published var showLyricMode = false
    @Published var showChildProfile = false
    @Published var showGames = false
    @Published var showDocumentation = false
    @Published var showModelSelector = false
    @Published var selectedPersonalityForChat: Personality?
    @Published var showSetupWizard = false
    @Published var isStartingDocker = false
    @Published var isOnline: Bool = false
    @Published var onlineTime: Date?
    @Published var learningTiles: [LearningTile] = []
    @Published var storyTiles: [StoryTile] = []

    private let personalityService: PersonalityServiceProtocol
    private let personalityRepository: PersonalityRepositoryProtocol
    private let tileRepository: TileRepositoryProtocol
    private var statusCheckTimer: Timer?

    init(personalityService: PersonalityServiceProtocol, personalityRepository: PersonalityRepositoryProtocol, tileRepository: TileRepositoryProtocol) {
        self.personalityService = personalityService
        self.personalityRepository = personalityRepository
        self.tileRepository = tileRepository
        loadPersonalities()
        loadTiles()
        checkFirstLaunch()
        startStatusMonitoring()
    }

    func checkFirstLaunch() {
        let hasCompletedSetup = UserDefaults.standard.bool(forKey: "hasCompletedSetup")
        if !hasCompletedSetup {
            showSetupWizard = true
        }
    }

    func startStatusMonitoring() {
        // Check status immediately
        Task {
            await checkMoxieStatus()
        }

        // Then check every 5 seconds
        statusCheckTimer = Timer.scheduledTimer(withTimeInterval: 5.0, repeats: true) { [weak self] _ in
            Task { @MainActor in
                await self?.checkMoxieStatus()
            }
        }
    }

    func checkMoxieStatus() async {
        let endpoint = UserDefaults.standard.string(forKey: "moxieEndpoint") ?? AppConfig.statusEndpoint

        guard let url = URL(string: endpoint) else {
            isOnline = false
            onlineTime = nil
            return
        }

        do {
            let (_, response) = try await URLSession.shared.data(from: url)
            if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 {
                if !isOnline {
                    // Just came online
                    onlineTime = Date()
                }
                isOnline = true
            } else {
                isOnline = false
                onlineTime = nil
            }
        } catch {
            isOnline = false
            onlineTime = nil
        }
    }

    deinit {
        statusCheckTimer?.invalidate()
    }

    func loadPersonalities() {
        let savedPersonalities = personalityRepository.loadPersonalities()
        allPersonalities = Personality.allPersonalities + savedPersonalities
        print("[ContentViewModel] Loaded \(allPersonalities.count) personalities: \(allPersonalities.map { $0.name })")
    }

    func loadTiles() {
        learningTiles = tileRepository.loadLearningTiles()
        storyTiles = tileRepository.loadStoryTiles()
    }

    func switchPersonality(_ personality: Personality) async {
        print("[ContentViewModel] switchPersonality called for: \(personality.name)")
        isUpdating = true
        statusMessage = "Switching to \(personality.name)..."

        do {
            print("[ContentViewModel] Calling personalityService...")
            try await personalityService.switchPersonality(personality)
            print("[ContentViewModel] Service call completed")
            statusMessage = "✅ SUCCESS! Moxie is now \(personality.emoji) \(personality.name)!"

            // Clear status after delay
            try? await Task.sleep(nanoseconds: 3_000_000_000)
            statusMessage = nil
        } catch {
            statusMessage = "❌ Error: \(error.localizedDescription)"
        }

        isUpdating = false
    }

    func editPersonality(_ personality: Personality) {
        editingPersonality = personality
        showPersonalityEditor = true
    }

    func openChat(for personality: Personality) {
        selectedPersonalityForChat = personality
        showChat = true
    }

    /// Find Docker executable path
    private var dockerPath: String {
        let paths = [
            "/usr/local/bin/docker",
            "/opt/homebrew/bin/docker",
            "/usr/bin/docker",
            "/Applications/Docker.app/Contents/Resources/bin/docker"
        ]
        return paths.first { FileManager.default.fileExists(atPath: $0) } ?? "docker"
    }

    /// Find docker-compose executable path
    private var dockerComposePath: String {
        let paths = [
            "/usr/local/bin/docker-compose",
            "/opt/homebrew/bin/docker-compose",
            "/usr/bin/docker-compose",
            "/Applications/Docker.app/Contents/Resources/bin/docker-compose"
        ]
        if let found = paths.first(where: { FileManager.default.fileExists(atPath: $0) }) {
            return found
        }
        // Fallback to docker compose (v2 syntax)
        return "\(dockerPath) compose"
    }

    func startDockerContainer() async {
        isStartingDocker = true
        statusMessage = "Starting OpenMoxie containers..."

        let openMoxieDir = FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent("OpenMoxie").path
        let docker = dockerPath
        let compose = dockerComposePath

        // Try to start existing containers first, otherwise use docker-compose
        let script = """
        # Try starting existing containers first
        if \(docker) start openmoxie-server openmoxie-mqtt 2>/dev/null; then
            echo "Started existing containers"
            exit 0
        fi

        # If that fails, try docker-compose in OpenMoxie directory
        if [ -d "\(openMoxieDir)" ] && [ -f "\(openMoxieDir)/docker-compose.yml" ]; then
            cd "\(openMoxieDir)" && \(compose) up -d
            exit $?
        fi

        # Last resort: pull and run the images directly
        \(docker) network create openmoxie-net 2>/dev/null || true

        \(docker) run -d --name openmoxie-mqtt \
          --network openmoxie-net \
          -p 1883:1883 \
          -p 8883:8883 \
          openmoxie/openmoxie-mqtt:latest

        \(docker) run -d --name openmoxie-server \
          --network openmoxie-net \
          -p 8001:8000 \
          openmoxie/openmoxie-server:latest
        """

        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/bin/bash")
        process.arguments = ["-c", script]

        let pipe = Pipe()
        process.standardOutput = pipe
        process.standardError = pipe

        do {
            try process.run()
            process.waitUntilExit()

            let data = pipe.fileHandleForReading.readDataToEndOfFile()
            let output = String(data: data, encoding: .utf8) ?? ""

            if process.terminationStatus == 0 {
                statusMessage = "✅ OpenMoxie containers started!"
                // Check status immediately
                await checkMoxieStatus()
            } else {
                statusMessage = "❌ Failed to start Docker: \(output)"
            }
        } catch {
            statusMessage = "❌ Error starting Docker: \(error.localizedDescription)"
        }

        // Clear status after delay
        try? await Task.sleep(nanoseconds: 5_000_000_000)
        statusMessage = nil
        isStartingDocker = false
    }
}